x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationIndexingService.java

```diff
@@ -0,0 +1,128 @@
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.index.IndexAction;
+import org.elasticsearch.action.index.IndexRequestBuilder;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.logging.DeprecatedLogHandler;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.logging.ESLogMessage;
+import org.elasticsearch.common.logging.LoggerMessageFormat;
+import org.elasticsearch.common.settings.Setting;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.elasticsearch.common.Strings.isNullOrEmpty;
+
+/**
+ * This service is responsible for writing deprecation messages to a data stream.
+ * The writing of messages can be toggled using the
+ * {@link #WRITE_DEPRECATION_LOGS_TO_INDEX} setting.
+ */
+public class DeprecationIndexingService extends AbstractLifecycleComponent implements ClusterStateListener, DeprecatedLogHandler {
+    private static final Logger LOGGER = LogManager.getLogger(DeprecationIndexingService.class);
+
+    private static final String DATA_STREAM_NAME = "logs-deprecation-elasticsearch";
+
+    private static final String DEPRECATION_ORIGIN = "deprecation";
+
+    public static final Setting<Boolean> WRITE_DEPRECATION_LOGS_TO_INDEX = Setting.boolSetting(
+        "cluster.deprecation_indexing.enabled",
+        false,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private final Client client;
+    private volatile boolean isEnabled = false;
+
+    public DeprecationIndexingService(ClusterService clusterService, Client client) {
+        this.client = new OriginSettingClient(client, DEPRECATION_ORIGIN);
+
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Indexes a deprecation message.
+     * @param key          the key that was used to determine if this deprecation should have been be logged.
+     *                     Useful when aggregating the recorded messages.
+     * @param esLogMessage the message to log
+     */
+    public void log(String key, String xOpaqueId, ESLogMessage esLogMessage) {
+        if (this.lifecycle.started() == false) {
+            return;
+        }
+
+        if (this.isEnabled == false) {
+            return;
+        }
+
+        String messagePattern = esLogMessage.getMessagePattern();
+        Object[] arguments = esLogMessage.getArguments();
+        String message = LoggerMessageFormat.format(messagePattern, arguments);
```

Comment:
not the biggest fan of that `LoggerMessageFormat` . It duplicates what log4j can do for us now - I am not familiar with history behind this though. At some point I think we should refactor that logic to use log4j.
Happy with this for now.

How about logger name or node/cluster name fields?